modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/TransactionStateResolver.java

```diff
@@ -44,29 +72,194 @@ public class TransactionStateResolver {
+            ReplicaService replicaService,
+            TxManager txManager,
+            HybridClock clock,
+            Function<String, ClusterNode> clusterNodeResolver,
+            Function<String, ClusterNode> clusterNodeResolverById,
+            Supplier<String> localNodeIdSupplier,
+            MessagingService messagingService
+    ) {
         this.replicaService = replicaService;
+        this.txManager = txManager;
+        this.clock = clock;
         this.clusterNodeResolver = clusterNodeResolver;
+        this.clusterNodeResolverById = clusterNodeResolverById;
+        this.localNodeId = new Lazy<>(localNodeIdSupplier);
+        this.messagingService = messagingService;
     }
 
     /**
-     * Sends a transaction state request to the primary replica.
+     * This should be called in order to allow the transaction state resolver to listen to {@link TxStateRequest} messages.
+     */
+    public void start() {
+        messagingService.addMessageHandler(TxMessageGroup.class, (msg, sender, correlationId) -> {
+            if (msg instanceof TxStateRequest) {
+                TxStateRequest req = (TxStateRequest) msg;
+
+                processTxStateRequest(req)
+                        .thenAccept(txStateMeta -> {
+                            NetworkMessage response = FACTORY.txStateResponse()
+                                    .txStateMeta(txStateMeta)
+                                    .timestampLong(clock.nowLong())
+                                    .build();
+
+                            messagingService.respond(sender, response, correlationId);
+                        });
+            }
+        });
+    }
+
+    /**
+     * Resolves transaction state locally, if possible, or distributively, if needed.
      *
-     * @param replicaGrp Replication group id.
-     * @param request Status request.
-     * @return Result future.
+     * @param txId Transaction id.
+     * @param commitGrpId Commit partition group id.
+     * @param timestamp Timestamp.
+     * @return Future with the transaction state meta as a result.
      */
-    public CompletableFuture<TxMeta> sendMetaRequest(ReplicationGroupId replicaGrp, TxStateReplicaRequest request) {
-        CompletableFuture<TxMeta> resFut = new CompletableFuture<>();
+    public CompletableFuture<TransactionMeta> resolveTxState(
+            UUID txId,
+            ReplicationGroupId commitGrpId,
+            HybridTimestamp timestamp
+    ) {
+        TxStateMeta localMeta = txManager.stateMeta(txId);
+
+        if (localMeta != null) {
+            if (isFinalState(localMeta.txState())) {
+                return completedFuture(localMeta);
+            }
+
+            // If the local node is a tx coordinator:
+            // if tx state is FINISHING, we will have to wait for the actual finish, otherwise we can return the current state.
+            if (localNodeId.get().equals(localMeta.txCoordinatorId()) && localMeta.txState() != FINISHING) {
```

Comment:
Don't worth to do it separately of common case, because if we have to send a message to local node the message bypass the network layer.